src/mongo/delegates/Views.class.php

```diff
@@ -532,6 +476,109 @@ public function generateView($viewId,$resource=null,$context=null,$queueName=nul
         }
     }
 
+    protected function ensureIndices($viewSpec, $collection) {
+
+        // ensure that the ID field, view type, and the impactIndex indexes are correctly set up
+        $collection->ensureIndex(
+            array(
+                '_id.r'=>1,
+                '_id.c'=>1,
+                '_id.type'=>1
+            ),
+            array(
+                'background'=>1
+            )
+        );
+
+        $collection->ensureIndex(
+            array(
+                '_id.type'=>1
+            ),
+            array(
+                'background'=>1
+            )
+        );
+
+        $collection->ensureIndex(
+            array(
+                'value.'._IMPACT_INDEX=>1
+            ),
+            array(
+                'background'=>1
+            )
+        );
+
+        // spec type/revision index lets us find outdated views
+        Config::ensureIndexForSpecRevision($collection);
+
+        // ensure any custom view indexes
+        if (isset($viewSpec['ensureIndexes']))
+        {
+            foreach ($viewSpec['ensureIndexes'] as $ensureIndex)
+            {
+                $collection->ensureIndex(
+                    $ensureIndex,
+                    array(
+                        'background'=>1
+                    )
+                );
+            }
+        }
+    }
+
+    /**
+    * Constructs a view document to save from the relevant root CBD document, and saves it.
+    *
+    * TODO: we must only save if $viewSpec[_revision] >= current saved view document spec.revision
+    * This will ensure that view specification revisions monotonically increase.
+    *
+    * @param array $viewSpec
+    * @param \MongoCollection $collection - the views composite collection
+    * @param array $doc - the root CBD from which this view is generated
+    * @param string $from - the name of the CBD collection from which $doc is drawn
+    * @param string $contextAlias - the context for this view
+    **/
+    public function saveGeneratedView($viewSpec, $collection, $doc, $from, $contextAlias) {
+        // set up ID
+        $generatedView = array(
+            "_id" => array(
+                _ID_RESOURCE=>$doc["_id"][_ID_RESOURCE],
+                _ID_CONTEXT=>$doc["_id"][_ID_CONTEXT],
+                _ID_TYPE=>$viewSpec['_id']
+            )
+        );
+
+        // if specification is revisioned, add revision information to what we will save
+        $specRevision = Config::getSpecRevisionSubDoc($viewSpec);
+        if(isset($specRevision)) {
+            $generatedView = array_merge($generatedView, $specRevision);
+        }
```

Comment:
We should only save this document if its revision is the greater-than-or-equal-to the revision of the existing document in the database.  If we do not ensure this, we can overwrite our document with copies generated by an old revision of the spec.  We must prevent that.